File: C++/InterviewQuestions/ZomatoFoodDeliveryDemo.cpp
/*
Zomato Food Delivery Demo
------------------------------

Functional Requirements:
    1. Search/Browse/Recommend by dish or by proximity to the customer's location
    2. Select the restaurant and explore it
    3. Add items to a cart
    4. Place order / make the payment / compute total
    5. Delivery partner assigned
    6. Track the order status

    ------ OUT OF SCOPE -----
    Support multiple types of payments
    Admin Service to manage the CRM

Non-Functional Requirements:
    No double assignment
    No lost/duplicate cart updates
    Valid state machine
    Extensibility

Entities:
    Customer
    FoodDeliveryService
    DeliveryPartner
    DeliveryPartnerAssignmentService
    RestrauntManagerService
    Restraunt, MenuItem
    Cart
    Order, OrderService
    PaymentGateway
*/

#include <cmath>
#include <iomanip>
#include <iostream>
#include <limits>
#include <map>
#include <memory>
#include <mutex>
#include <stdexcept>
#include <string>
#include <vector>

static int nextId = 1;

std::string generateId(const std::string& prefix) {
    return prefix + std::to_string(nextId++);
}

struct Location {
    double x;
    double y;

    double distanceTo(const Location& other) const {
        return std::sqrt(std::pow(x - other.x, 2) + std::pow(y - other.y, 2));
    }
};

struct MenuItem {
    std::string id;
    std::string name;
    double price;
    bool available;

    MenuItem(std::string name, double price)
        : id(generateId("item-")), name(std::move(name)), price(price), available(true) {}
};

class Restraunt {
public:
    std::string id;
    std::string name;
    Location location;
    bool open = true;

    Restraunt(std::string name, Location location)
        : id(generateId("rest-")), name(std::move(name)), location(location) {}

    void addMenuItem(std::shared_ptr<MenuItem> item) { menu[item->id] = item; }

    std::vector<std::shared_ptr<MenuItem>> getMenu() const {
        std::vector<std::shared_ptr<MenuItem>> items;
        for (auto& [id, item] : menu) items.push_back(item);
        return items;
    }

    bool hasDish(const std::string& dishName) const {
        for (auto& [id, item] : menu) {
            if (item->available && item->name == dishName) return true;
        }
        return false;
    }

private:
    std::map<std::string, std::shared_ptr<MenuItem>> menu;
};

class RestrauntManagerService {
public:
    std::shared_ptr<Restraunt> addRestraunt(std::shared_ptr<Restraunt> restraunt) {
        restraunts[restraunt->id] = restraunt;
        return restraunt;
    }

    std::vector<std::shared_ptr<Restraunt>> searchByDish(const std::string& dishName) const {
        std::vector<std::shared_ptr<Restraunt>> matches;
        for (auto& [id, r] : restraunts) {
            if (r->open && r->hasDish(dishName)) matches.push_back(r);
        }
        return matches;
    }

    std::vector<std::shared_ptr<Restraunt>> searchByProximity(const Location& customerLocation, double radius) const {
        std::vector<std::shared_ptr<Restraunt>> matches;
        for (auto& [id, r] : restraunts) {
            if (r->open && r->location.distanceTo(customerLocation) <= radius) matches.push_back(r);
        }
        return matches;
    }

private:
    std::map<std::string, std::shared_ptr<Restraunt>> restraunts;
};

class Cart {
public:
    mutable std::recursive_mutex lock;
    std::string restrauntId;
    std::map<std::shared_ptr<MenuItem>, int> items;

    void addItem(const std::shared_ptr<Restraunt>& restraunt, const std::shared_ptr<MenuItem>& item, int quantity) {
        std::lock_guard<std::recursive_mutex> guard(lock);
        if (!restrauntId.empty() && restrauntId != restraunt->id) {
            throw std::runtime_error("Cart already has items from another restraunt; clear it first");
        }
        restrauntId = restraunt->id;
        items[item] += quantity;
    }

    double total() const {
        std::lock_guard<std::recursive_mutex> guard(lock);
        double sum = 0;
        for (auto& [item, qty] : items) sum += item->price * qty;
        return sum;
    }

    void clear() {
        std::lock_guard<std::recursive_mutex> guard(lock);
        items.clear();
        restrauntId.clear();
    }
};

class Customer {
public:
    std::string id;
    std::string name;
    Location address;
    Cart cart;

    Customer(std::string name, Location address)
        : id(generateId("cust-")), name(std::move(name)), address(address) {}
};

class DeliveryPartner {
public:
    std::string id;
    std::string name;
    Location currentLocation;
    bool available = true;

    DeliveryPartner(std::string name, Location currentLocation)
        : id(generateId("partner-")), name(std::move(name)), currentLocation(currentLocation) {}
};

struct PaymentGateway {
    virtual bool pay(double amount) = 0;
    virtual ~PaymentGateway() = default;
};

class MockPaymentGateway : public PaymentGateway {
public:
    bool pay(double amount) override {
        std::cout << std::fixed << std::setprecision(2);
        std::cout << "Charged " << amount << " via mock payment gateway" << std::endl;
        return true;
    }
};

enum class OrderStatus { PLACED, PARTNER_ASSIGNED, PICKED_UP, DELIVERED, CANCELLED };

std::string orderStatusName(OrderStatus status) {
    switch (status) {
        case OrderStatus::PLACED: return "PLACED";
        case OrderStatus::PARTNER_ASSIGNED: return "PARTNER_ASSIGNED";
        case OrderStatus::PICKED_UP: return "PICKED_UP";
        case OrderStatus::DELIVERED: return "DELIVERED";
        case OrderStatus::CANCELLED: return "CANCELLED";
    }
    return "UNKNOWN";
}

class Order {
public:
    std::string id;
    std::shared_ptr<Customer> customer;
    std::shared_ptr<Restraunt> restraunt;
    std::map<std::shared_ptr<MenuItem>, int> items;
    double totalAmount;
    OrderStatus status = OrderStatus::PLACED;
    std::shared_ptr<DeliveryPartner> deliveryPartner;

    Order(std::shared_ptr<Customer> customer, std::shared_ptr<Restraunt> restraunt,
          std::map<std::shared_ptr<MenuItem>, int> items, double totalAmount)
        : id(generateId("order-")), customer(std::move(customer)), restraunt(std::move(restraunt)),
          items(std::move(items)), totalAmount(totalAmount) {}
};

// Assigns the nearest available delivery partner to the restraunt; no partner
// is ever handed two active orders at once, which rules out double assignment.
class DeliveryPartnerAssignmentService {
public:
    void registerPartner(std::shared_ptr<DeliveryPartner> partner) { partners.push_back(std::move(partner)); }

    std::shared_ptr<DeliveryPartner> assign(const std::shared_ptr<Order>& order) {
        std::lock_guard<std::mutex> guard(mutex);
        std::shared_ptr<DeliveryPartner> nearest;
        double bestDistance = std::numeric_limits<double>::max();
        for (auto& partner : partners) {
            if (!partner->available) continue;
            double distance = partner->currentLocation.distanceTo(order->restraunt->location);
            if (distance < bestDistance) {
                bestDistance = distance;
                nearest = partner;
            }
        }
        if (nearest) nearest->available = false;
        return nearest;
    }

    void release(const std::shared_ptr<DeliveryPartner>& partner) {
        std::lock_guard<std::mutex> guard(mutex);
        partner->available = true;
    }

private:
    std::vector<std::shared_ptr<DeliveryPartner>> partners;
    std::mutex mutex;
};

class OrderService {
public:
    OrderService(std::shared_ptr<PaymentGateway> paymentGateway,
                 std::shared_ptr<DeliveryPartnerAssignmentService> assignmentService)
        : paymentGateway(std::move(paymentGateway)), assignmentService(std::move(assignmentService)) {}

    std::shared_ptr<Order> placeOrder(std::shared_ptr<Customer> customer, std::shared_ptr<Restraunt> restraunt) {
        Cart& cart = customer->cart;
        std::shared_ptr<Order> order;
        {
            std::lock_guard<std::recursive_mutex> guard(cart.lock);
            if (cart.items.empty()) {
                throw std::runtime_error("Cannot place an order with an empty cart");
            }
            order = std::make_shared<Order>(customer, restraunt, cart.items, cart.total());
            orders[order->id] = order;
            cart.items.clear();
            cart.restrauntId.clear();
        }

        paymentGateway->pay(order->totalAmount);

        auto partner = assignmentService->assign(order);
        if (partner) {
            order->deliveryPartner = partner;
            order->status = OrderStatus::PARTNER_ASSIGNED;
        }
        return order;
    }

    void advanceStatus(const std::string& orderId, OrderStatus status) {
        auto it = orders.find(orderId);
        if (it == orders.end()) throw std::invalid_argument("Unknown order: " + orderId);
        it->second->status = status;
        if (status == OrderStatus::DELIVERED && it->second->deliveryPartner) {
            assignmentService->release(it->second->deliveryPartner);
        }
    }

    std::shared_ptr<Order> trackOrder(const std::string& orderId) {
        auto it = orders.find(orderId);
        return it == orders.end() ? nullptr : it->second;
    }

private:
    std::shared_ptr<PaymentGateway> paymentGateway;
    std::shared_ptr<DeliveryPartnerAssignmentService> assignmentService;
    std::map<std::string, std::shared_ptr<Order>> orders;
};

class FoodDeliveryService {
public:
    RestrauntManagerService restrauntManagerService;
    std::shared_ptr<DeliveryPartnerAssignmentService> assignmentService =
        std::make_shared<DeliveryPartnerAssignmentService>();
    OrderService orderService{std::make_shared<MockPaymentGateway>(), assignmentService};
};

int main() {
    FoodDeliveryService service;

    auto tasty = service.restrauntManagerService.addRestraunt(
        std::make_shared<Restraunt>("Tasty Bites", Location{0, 0}));
    auto paneerTikka = std::make_shared<MenuItem>("Paneer Tikka", 220.0);
    auto naan = std::make_shared<MenuItem>("Butter Naan", 40.0);
    tasty->addMenuItem(paneerTikka);
    tasty->addMenuItem(naan);

    auto spice = service.restrauntManagerService.addRestraunt(
        std::make_shared<Restraunt>("Spice Route", Location{5, 5}));
    spice->addMenuItem(std::make_shared<MenuItem>("Chicken Biryani", 260.0));

    service.assignmentService->registerPartner(std::make_shared<DeliveryPartner>("Ravi", Location{0.5, 0.5}));
    service.assignmentService->registerPartner(std::make_shared<DeliveryPartner>("Neha", Location{4.5, 4.5}));

    auto alice = std::make_shared<Customer>("Alice", Location{1, 1});

    std::cout << "Search by dish 'Paneer Tikka': " << std::endl;
    for (auto& r : service.restrauntManagerService.searchByDish("Paneer Tikka")) {
        std::cout << "  - " << r->name << std::endl;
    }

    std::cout << "Search by proximity (radius 3 from Alice):" << std::endl;
    for (auto& r : service.restrauntManagerService.searchByProximity(alice->address, 3)) {
        std::cout << "  - " << r->name << std::endl;
    }

    alice->cart.addItem(tasty, paneerTikka, 1);
    alice->cart.addItem(tasty, naan, 2);
    std::cout << std::fixed << std::setprecision(2);
    std::cout << "Cart total: " << alice->cart.total() << std::endl;

    auto order = service.orderService.placeOrder(alice, tasty);
    std::cout << "Order placed: " << order->id << " status=" << orderStatusName(order->status)
              << " partner=" << (order->deliveryPartner ? order->deliveryPartner->name : "none") << std::endl;

    service.orderService.advanceStatus(order->id, OrderStatus::PICKED_UP);
    std::cout << "Order status: " << orderStatusName(service.orderService.trackOrder(order->id)->status) << std::endl;

    service.orderService.advanceStatus(order->id, OrderStatus::DELIVERED);
    std::cout << "Order status: " << orderStatusName(service.orderService.trackOrder(order->id)->status) << std::endl;

    return 0;
}

File: Python/InterviewQuestions/ZomatoFoodDeliveryDemo.py
"""
Zomato Food Delivery Demo
------------------------------

Functional Requirements:
    1. Search/Browse/Recommend by dish or by proximity to the customer's location
    2. Select the restaurant and explore it
    3. Add items to a cart
    4. Place order / make the payment / compute total
    5. Delivery partner assigned
    6. Track the order status

    ------ OUT OF SCOPE -----
    Support multiple types of payments
    Admin Service to manage the CRM

Non-Functional Requirements:
    No double assignment
    No lost/duplicate cart updates
    Valid state machine
    Extensibility

Entities:
    Customer
    FoodDeliveryService
    DeliveryPartner
    DeliveryPartnerAssignmentService
    RestrauntManagerService
    Restraunt, MenuItem
    Cart
    Order, OrderService
    PaymentGateway
"""

import math
import threading
import uuid
from abc import ABC, abstractmethod
from enum import Enum, auto
from typing import Dict, List, Optional


class Location:
    def __init__(self, x: float, y: float):
        self.x = x
        self.y = y

    def distance_to(self, other: "Location") -> float:
        return math.sqrt((self.x - other.x) ** 2 + (self.y - other.y) ** 2)


class MenuItem:
    def __init__(self, name: str, price: float):
        self.id = str(uuid.uuid4())
        self.name = name
        self.price = price
        self.available = True


class Restraunt:
    def __init__(self, name: str, location: Location):
        self.id = str(uuid.uuid4())
        self.name = name
        self.location = location
        self.menu: Dict[str, MenuItem] = {}
        self.open = True

    def add_menu_item(self, item: MenuItem) -> None:
        self.menu[item.id] = item

    def get_menu(self) -> List[MenuItem]:
        return list(self.menu.values())

    def has_dish(self, dish_name: str) -> bool:
        return any(item.available and item.name.lower() == dish_name.lower() for item in self.menu.values())


class RestrauntManagerService:
    def __init__(self):
        self._restraunts: Dict[str, Restraunt] = {}

    def add_restraunt(self, restraunt: Restraunt) -> Restraunt:
        self._restraunts[restraunt.id] = restraunt
        return restraunt

    def search_by_dish(self, dish_name: str) -> List[Restraunt]:
        return [r for r in self._restraunts.values() if r.open and r.has_dish(dish_name)]

    def search_by_proximity(self, customer_location: Location, radius: float) -> List[Restraunt]:
        return [
            r for r in self._restraunts.values()
            if r.open and r.location.distance_to(customer_location) <= radius
        ]


class Cart:
    def __init__(self):
        self.lock = threading.RLock()
        self.restraunt_id: Optional[str] = None
        self.items: Dict[MenuItem, int] = {}

    def add_item(self, restraunt: Restraunt, item: MenuItem, quantity: int) -> None:
        with self.lock:
            if self.restraunt_id is not None and self.restraunt_id != restraunt.id:
                raise ValueError("Cart already has items from another restraunt; clear it first")
            self.restraunt_id = restraunt.id
            self.items[item] = self.items.get(item, 0) + quantity

    def total(self) -> float:
        with self.lock:
            return sum(item.price * qty for item, qty in self.items.items())

    def clear(self) -> None:
        with self.lock:
            self.items.clear()
            self.restraunt_id = None


class Customer:
    def __init__(self, name: str, address: Location):
        self.id = str(uuid.uuid4())
        self.name = name
        self.address = address
        self.cart = Cart()


class DeliveryPartner:
    def __init__(self, name: str, current_location: Location):
        self.id = str(uuid.uuid4())
        self.name = name
        self.current_location = current_location
        self.available = True


class PaymentGateway(ABC):
    @abstractmethod
    def pay(self, amount: float) -> bool:
        pass


class MockPaymentGateway(PaymentGateway):
    def pay(self, amount: float) -> bool:
        print(f"Charged {amount:.2f} via mock payment gateway")
        return True


class OrderStatus(Enum):
    PLACED = auto()
    PARTNER_ASSIGNED = auto()
    PICKED_UP = auto()
    DELIVERED = auto()
    CANCELLED = auto()


class Order:
    def __init__(self, customer: Customer, restraunt: Restraunt, items: Dict[MenuItem, int], total_amount: float):
        self.id = str(uuid.uuid4())
        self.customer = customer
        self.restraunt = restraunt
        self.items = items
        self.total_amount = total_amount
        self.status = OrderStatus.PLACED
        self.delivery_partner: Optional[DeliveryPartner] = None


class DeliveryPartnerAssignmentService:
    """Assigns the nearest available delivery partner to the restraunt; no
    partner is ever handed two active orders at once, which rules out
    double assignment."""

    def __init__(self):
        self._partners: List[DeliveryPartner] = []
        self._lock = threading.Lock()

    def register(self, partner: DeliveryPartner) -> None:
        self._partners.append(partner)

    def assign(self, order: Order) -> Optional[DeliveryPartner]:
        with self._lock:
            nearest = None
            best_distance = float("inf")
            for partner in self._partners:
                if not partner.available:
                    continue
                distance = partner.current_location.distance_to(order.restraunt.location)
                if distance < best_distance:
                    best_distance = distance
                    nearest = partner
            if nearest is not None:
                nearest.available = False
            return nearest

    def release(self, partner: DeliveryPartner) -> None:
        with self._lock:
            partner.available = True


class OrderService:
    def __init__(self, payment_gateway: PaymentGateway, assignment_service: DeliveryPartnerAssignmentService):
        self._payment_gateway = payment_gateway
        self._assignment_service = assignment_service
        self._orders: Dict[str, Order] = {}

    def place_order(self, customer: Customer, restraunt: Restraunt) -> Order:
        cart = customer.cart
        with cart.lock:
            if not cart.items:
                raise ValueError("Cannot place an order with an empty cart")
            order = Order(customer, restraunt, dict(cart.items), cart.total())
            self._orders[order.id] = order
            cart.items.clear()
            cart.restraunt_id = None

        self._payment_gateway.pay(order.total_amount)

        partner = self._assignment_service.assign(order)
        if partner is not None:
            order.delivery_partner = partner
            order.status = OrderStatus.PARTNER_ASSIGNED
        return order

    def advance_status(self, order_id: str, status: OrderStatus) -> None:
        order = self._orders.get(order_id)
        if order is None:
            raise ValueError(f"Unknown order: {order_id}")
        order.status = status
        if status == OrderStatus.DELIVERED and order.delivery_partner is not None:
            self._assignment_service.release(order.delivery_partner)

    def track_order(self, order_id: str) -> Optional[Order]:
        return self._orders.get(order_id)


class FoodDeliveryService:
    def __init__(self):
        self.restraunt_manager_service = RestrauntManagerService()
        self.assignment_service = DeliveryPartnerAssignmentService()
        self.order_service = OrderService(MockPaymentGateway(), self.assignment_service)


def main() -> None:
    service = FoodDeliveryService()

    tasty = service.restraunt_manager_service.add_restraunt(Restraunt("Tasty Bites", Location(0, 0)))
    paneer_tikka = MenuItem("Paneer Tikka", 220.0)
    naan = MenuItem("Butter Naan", 40.0)
    tasty.add_menu_item(paneer_tikka)
    tasty.add_menu_item(naan)

    spice = service.restraunt_manager_service.add_restraunt(Restraunt("Spice Route", Location(5, 5)))
    spice.add_menu_item(MenuItem("Chicken Biryani", 260.0))

    service.assignment_service.register(DeliveryPartner("Ravi", Location(0.5, 0.5)))
    service.assignment_service.register(DeliveryPartner("Neha", Location(4.5, 4.5)))

    alice = Customer("Alice", Location(1, 1))

    print("Search by dish 'Paneer Tikka': ")
    for r in service.restraunt_manager_service.search_by_dish("Paneer Tikka"):
        print(f"  - {r.name}")

    print("Search by proximity (radius 3 from Alice):")
    for r in service.restraunt_manager_service.search_by_proximity(alice.address, 3):
        print(f"  - {r.name}")

    alice.cart.add_item(tasty, paneer_tikka, 1)
    alice.cart.add_item(tasty, naan, 2)
    print(f"Cart total: {alice.cart.total():.2f}")

    order = service.order_service.place_order(alice, tasty)
    partner_name = order.delivery_partner.name if order.delivery_partner else "none"
    print(f"Order placed: {order.id} status={order.status.name} partner={partner_name}")

    service.order_service.advance_status(order.id, OrderStatus.PICKED_UP)
    print(f"Order status: {service.order_service.track_order(order.id).status.name}")

    service.order_service.advance_status(order.id, OrderStatus.DELIVERED)
    print(f"Order status: {service.order_service.track_order(order.id).status.name}")


if __name__ == "__main__":
    main()
